### ISSUE TITLE
Freefit drastically increases VRAM usage vs the previous token bucket bands

### ISSUE THREAD

**Issue by 2Raven2** [sorryhyun/anima_lora#58]
Before v1.11.2, I was able to train a 64 DIM LoRA at 1536^2 on an RTX 4090 (batch 1, 2 gradient accumulation steps), but after this update forced the use of freefit, it has massively increased VRAM costs by a further north of ~6 GB and now would need an RTX 5090 minimum if you want to train fast without using gradient checkpointing. So now it went up from ~23 GB to ~29-30 GB.

Is there a reason on the backend side for the previous token bucket bands to not co-exist with freefit? I'm aware that it may be superior, but seeing as the point of this trainer was low VRAM costs and speed, freefit seems to go against that.

I've had no problems so far with the previous implementation, so as it currently stands I'll downgrade to v1.11.1. 
Thanks for your work though!

**Comment by sorryhyun** [comment]
Hello, thanks for raising the issue. Before v1.11.1., 1536^2 res tier used single 8640 tier which cropped every image to fit on following resolutions:

`1440*1536, 1280*1728, 1152*1920, 1024*2160, ... etc`

So compilation was static and resolved single token (8640) which contributed lowering ram. Besides there was custom autograd function which was excluded from compilation boundary so it lowered about 10~20% lower vram.

However now we use about 10%+- token interval to preserve original ratio so almost no images are cropped. hence, actual token usage in 1536^2 tier may reach about 8836 tokens at maximum.

That said, we implemented dynamic shape compilation which also allows to lever vram usage by `activation_memory_budget`. Try using `activation_memory_budget=0.75` so it will make peak vram much lower.

**Comment by 2Raven2** [comment]
I've just tried it, 0.75 was still far too high, 0.55 was finally enough to fit into VRAM. However, because of the very low budget now, it's only a tiny bit faster than gradient checkpointing which lowered overall VRAM cost to 11 GB, compared to ~23 GB with `activation_memory_budget=0.55` while only saving 30 minutes.

Total time to train LoRA with `activation_memory_budget=0.55`: 8 hrs 45 mins (~23 GB VRAM required)
Total time to train LoRA with `gradient_checkpointing`: 9 hrs 15 mins (~11 GB VRAM required)
Total time to train LoRA without `gradient_checkpointing` and full activation budget in v1.11.1: 6hrs 30 mins

The LoRA requires a very low LR and can't be aggressively trained as much as I would prefer to save lots of time, it needs to be trained very slowly beyond 10,000 steps. I've tried a lot of LR between 1e-5 and 1e-4 and using cosine with restarts with varying aggressiveness; 3e-5 has been the best for this very niche concept I'm training for, so this is why I need a lot of steps for it to work.

More than double the VRAM required to save only 30 mins, at that point I'd just use gradient checkpointing.
This update has made high-res training ~50% slower for me with the RTX 4090.

**Comment by sorryhyun** [comment]
I see, sorry for that. I could not test 1536^2 resolution tier since what I have is 5070ti 16gb garbage. Anyway a core part of the problems is custom autograd which halved vram by excluding compile region. I'll check would there be some ways to avoid this issue. Sorry again for inconvenience without notifications.

**Comment by 2Raven2** [comment]
I did further LoRA training on the newer versions and the safe values for `activation_memory_budget` is between `0.45` and `0.50`. I've included some screenshots in case they may help, but let me know if you need more info:

<img width="1065" height="440" alt="Image" src="https://github.com/user-attachments/assets/0a7d03c8-a8c0-4b23-96b3-83c3c076a17d" />

<img width="2373" height="966" alt="Image" src="https://github.com/user-attachments/assets/1cb1298b-dfa5-42f8-9c9e-df6d4fe4a25f" />

No problems if there's ultimately no workarounds to this.

**Comment by sorryhyun** [comment]
hmm.... I see. what happens if you use `partitioner_aggressive_recomputation=true` the feature added in recent commit?
